common/meta/api/src/schema_api_impl.rs

```diff
@@ -227,18 +389,83 @@ impl<KV: KVApi> SchemaApi for KV {
             let (db_id_seq, _db_id) = get_id_value(self, &tenant_newdbname).await?;
             db_has_to_not_exist(db_id_seq, &tenant_newdbname, "rename_database")?;
 
+            // get db id list from _fd_db_id_list/<tenant>/<db_name>
+            let dbid_idlist = DbIdListKey {
+                tenant: tenant_dbname.tenant.clone(),
+                db_name: tenant_dbname.db_name.clone(),
+            };
+            let (db_id_list_seq, db_id_list_opt): (_, Option<DbIdList>) =
+                get_struct_value(self, &dbid_idlist).await?;
+            let mut db_id_list: DbIdList;
+            if db_id_list_seq == 0 {
+                db_id_list = DbIdList::new();
+                db_id_list.append(old_db_id);
+            } else {
+                match db_id_list_opt {
+                    Some(list) => db_id_list = list,
+                    None => {
+                        db_id_list = DbIdList::new();
+                        db_id_list.append(old_db_id);
+                    }
+                }
+            };
+
+            if let Some(last_db_id) = db_id_list.last() {
+                if *last_db_id != old_db_id {
+                    return Err(MetaError::AppError(AppError::DatabaseAlreadyExists(
+                        DatabaseAlreadyExists::new(
+                            &tenant_dbname.db_name,
+                            format!("rename_database: {} with a wrong db id", tenant_dbname),
+                        ),
+                    )));
+                }
+            } else {
+                return Err(MetaError::AppError(AppError::DatabaseAlreadyExists(
+                    DatabaseAlreadyExists::new(
+                        &tenant_dbname.db_name,
+                        format!("rename_database: {} with none db id history", tenant_dbname),
+                    ),
+                )));
+            }
+
+            let new_dbid_idlist = DbIdListKey {
+                tenant: tenant_dbname.tenant.clone(),
+                db_name: req.new_db_name.clone(),
+            };
+            let (new_db_id_list_seq, new_db_id_list_opt): (_, Option<DbIdList>) =
+                get_struct_value(self, &new_dbid_idlist).await?;
+            let mut new_db_id_list: DbIdList;
+            if new_db_id_list_seq == 0 {
+                new_db_id_list = DbIdList::new();
+            } else {
+                match new_db_id_list_opt {
+                    Some(list) => new_db_id_list = list,
+                    None => {
+                        new_db_id_list = DbIdList::new();
+                    }
+                }
+            };
+
             // rename database
             {
+                // move db id from old db id list to new db id list
+                db_id_list.pop();
+                new_db_id_list.append(old_db_id);
```

Comment:
`db_id_list` feels like a hash-map, but not a list, if it is allowed to undrop a db whose id is not at the top in `db_id_list`.

Reply:
currently, undrop only the top in `db_id_list`.